### PR TITLE
plugins/vim-illuminate: init + test

### PR DESCRIPTION
**Pull request by anriha**
Add support for [vim-illuminate](https://github.com/RRethy/vim-illuminate).

### CODE REVIEW

File: plugins/utils/vim-illuminate.nix [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  ...
+}:
+with lib; let
+  cfg = config.plugins.vim-illuminate;
+  helpers = import ../helpers.nix {inherit lib;};
+
+  mkListStr = helpers.defaultNullOpts.mkNullable (types.listOf types.str);
+
+  commonOptions = with helpers.defaultNullOpts; {
+    providers = mkListStr ''[ "lsp" "treesitter" "regex" ]'' ''
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:14]:
```suggestion
    providers = mkListStr ''["lsp" "treesitter" "regex"]'' ''
```

File: plugins/utils/vim-illuminate.nix [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      Provider used to get references in the buffer, ordered by priority
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:15]:
```suggestion
      Provider used to get references in the buffer, ordered by priority.
```

File: plugins/utils/vim-illuminate.nix [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    '';
+
+    delay = mkInt 100 ''
+      Delay in milliseconds
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:19]:
```suggestion
      Delay in milliseconds.
```

File: plugins/utils/vim-illuminate.nix [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    '';
+
+    modesDenylist = mkListStr "[]" ''
+      Modes to not illuminate, this overrides modes_allowlist
+      See `:help mode()` for possible values
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:25]:
```suggestion
    modesDenylist = mkListStr "[]" ''
      Modes to not illuminate, this overrides `modes_allowlist`.
      See `:help mode()` for possible values.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    modesAllowlist = mkListStr "[]" ''
+      Modes to illuminate, this is overriden by modes_denylist
+      See `:help mode()` for possible values
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:30]:
```suggestion
    modesAllowlist = mkListStr "[]" ''
      Modes to illuminate, this is overriden by `modes_denylist`.
      See `:help mode()` for possible values.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    providersRegexSyntaxDenylist = mkListStr "[]" ''
+      Syntax to not illuminate, this overrides providers_regex_syntax_allowlist
+      Only applies to the 'regex' provider
+      Use :echom synIDattr(synIDtrans(synID(line('.'), col('.'), 1)), 'name')
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:36]:
```suggestion
    providersRegexSyntaxDenylist = mkListStr "[]" ''
      Syntax to not illuminate, this overrides `providers_regex_syntax_allowlist`.
      Only applies to the 'regex' provider.
      Use `:echo synIDattr(synIDtrans(synID(line('.'), col('.'), 1)), 'name')`.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    providersRegexSyntaxAllowlist = mkListStr "[]" ''
+      Syntax to illuminate, this is overriden by providers_regex_syntax_denylist
+      Only applies to the 'regex' provider
+      Use :echom synIDattr(synIDtrans(synID(line('.'), col('.'), 1)), 'name')
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:42]:
```suggestion
    providersRegexSyntaxAllowlist = mkListStr "[]" ''
      Syntax to illuminate, this is overriden by `providers_regex_syntax_denylist`.
      Only applies to the 'regex' provider.
      Use `:echo synIDattr(synIDtrans(synID(line('.'), col('.'), 1)), 'name')`.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    underCursor = mkBool true ''
+      Whether or not to illuminate under the cursor
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:45]:
```suggestion
      Whether or not to illuminate under the cursor.
```

File: plugins/utils/vim-illuminate.nix [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    '';
+
+    largeFileCutoff = helpers.mkNullOrOption types.int ''
+      Number of lines at which to use large_file_config
+      The `under_cursor` option is disabled when this cutoff is hit
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:51]:
```suggestion
    largeFileCutoff = helpers.mkNullOrOption types.int ''
      Number of lines at which to use `large_file_config`.
      The `under_cursor` option is disabled when this cutoff is hit.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+    minCountToHighlight = mkInt 1 ''
+      Minimum number of matches required to perform highlighting
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:54]:
```suggestion
      Minimum number of matches required to perform highlighting.
```

File: plugins/utils/vim-illuminate.nix [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    '';
+  };
+
+  filetypeOptions = {
+    filetypesDenylist = mkListStr ''[ "dirvish" "fugitive" ]'' ''
+      Filetypes to not illuminate, this overrides filetypes_allowlist
+    '';
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:61]:
```suggestion
    filetypesDenylist = mkListStr ''["dirvish" "fugitive"]'' ''
      Filetypes to not illuminate, this overrides `filetypes_allowlist`.
    '';
```

File: plugins/utils/vim-illuminate.nix [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+    filetypesAllowlist = mkListStr "[]" ''
+      Filetypes to illuminate, this is overriden by filetypes_denylist
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:64]:
```suggestion
      Filetypes to illuminate, this is overriden by `filetypes_denylist`.
```

File: plugins/utils/vim-illuminate.nix [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+      {
+        large_file_overrides = commonSetupOptions largeFileOverrides // filetypeSetupOptions largeFileOverrides;
+
+        filetype_overrides = let
+          override = attr: {${attr.filetype} = commonSetupOptions attr.overrides;};
+        in
+          map override filetypeOverrides;
+      }
+      // filetypeSetupOptions cfg
+      // commonSetupOptions cfg;
```
Comment (GaetanLepage) [inline plugins/utils/vim-illuminate.nix:135]:
```suggestion
        large_file_overrides = (commonSetupOptions largeFileOverrides) // (filetypeSetupOptions largeFileOverrides);

        filetype_overrides = let
          override = attr: {${attr.filetype} = commonSetupOptions attr.overrides;};
        in
          map override filetypeOverrides;
      }
      // (filetypeSetupOptions cfg)
      // (commonSetupOptions cfg);
```

This one is only a suggestion. Those parenthesizes are not mandatory but help with readability (I think).

File: tests/test-sources/plugins/utils/vim-illuminate.nix [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+          };
+        }
+      ];
+
+      largeFileOverrides = {
+        delay = 20;
+        underCursor = true;
+      };
+    };
+  };
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/utils/vim-illuminate.nix:34]:
```suggestion
      enable = true;

      delay = 50;
      providers = ["lsp"];
      modesDenylist = ["n"];
      modesAllowlist = ["v"];
      underCursor = false;
      largeFileCutoff = 10;
      minCountToHighlight = 2;
      filetypesDenylist = ["csharp"];
      filetypesAllowlist = ["python"];
      filetypeOverrides = [
        {
          filetype = "c";
          overrides = {
            delay = 10;
            providers = ["treesitter"];
          };
        }
      ];
      largeFileOverrides = {
        delay = 20;
        underCursor = true;
      };
    };
  };
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Thanks for this nice contribution !
I just have a few formatting change requests. Apart from this, it looks good.

**Comment by GaetanLepage** [comment]
Also, I would rather see this plugin named `illuminate` instead of `vim-illuminate`.

**Comment by anriha** [comment]
> Also, I would rather see this plugin named `illuminate` instead of `vim-illuminate`.

What is the guideline for this? Just use name that you use in lua require?

**Comment by GaetanLepage** [comment]
> What is the guideline for this? Just use name that you use in lua require?

Unfortunately we do not have a strong and deterministic guideline on the naming.
The current codebase already has some discrepancies in the naming of plugins.

If there is no clash or other exception, I tend to prefer sticking to the _namespace name_ (i.e. `telescope`, `harpoon`, `treesitter`...).
So in this case, I would go for `illuminate`.

**Comment by anriha** [comment]
> > What is the guideline for this? Just use name that you use in lua require?
> 
> Unfortunately we do not have a strong and deterministic guideline on the naming. The current codebase already has some discrepancies in the naming of plugins.
> 
> If there is no clash or other exception, I tend to prefer sticking to the _namespace name_ (i.e. `telescope`, `harpoon`, `treesitter`...). So in this case, I would go for `illuminate`.

Ok, makes sense, renamed :+1:

File: plugins/utils/illuminate.nix [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    '';
+  };
+in {
+  options.plugins.illuminate = with helpers;
+  with defaultNullOpts;
+    helpers.extraOptionsOptions
+    // {
+      enable = mkEnableOption "illuminate";
+
+      package = mkPackageOption "illuminate" pkgs.vimPlugins.vim-illuminate;
```
Comment (GaetanLepage) [inline plugins/utils/illuminate.nix:74]:
```suggestion
      enable = mkEnableOption "vim-illuminate";

      package = mkPackageOption "vim-illuminate" pkgs.vimPlugins.vim-illuminate;
```

Maybe just here, we can leave the real plugin's name.

Comment (GaetanLepage) [reply plugins/utils/illuminate.nix:74]: Well even the github page of the plugin uses both `vim-illuminate` and `illuminate.vim` X)

Comment (anriha) [reply plugins/utils/illuminate.nix:74]: Never really understood why someone uses .vim -nvim or .nvim makes at least some sense, but this is weird, especially when it is mainly neovim plugin. I think it still makes sense to change those two, so they match nixpkgs.

**Review by GaetanLepage — APPROVED** [review]

**Pull request merged**